---
author: "Centre for Population Genomics"
date: "`r Sys.time()`"
output:
  html_document:
    theme: cosmo
    css: style.css
    toc: true
    code_download: true
    code_folding: hide
  rmdformats::material:
    highlight: kate
params:
  title: "ExpansionHunter REViewer Results"
  variant_catalog: "expansionhunter/variant_catalog.json"
  indir: "expansionhunter/output"
description: "ExpansionHunter REViewer Results"
title: "`r params$title`"
---

```{r knitr_opts, include=F}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r load_pkgs}
require(fs, include.only = "dir_ls")
require(glue)
require(here)
require(jsonlite, include.only = "read_json")
require(kableExtra, include.only = c("kable_minimal", "kable_material"))
require(knitr)
require(tidyverse)
```

```{r read_params, eval=FALSE, echo=FALSE}
# for interactive debugging
params <- rmarkdown::yaml_front_matter(here("inst/rmd/str/expansionhunter_reviewer.Rmd"))$params
```

## Introduction

Below we show the main output from
[ExpansionHunter](https://github.com/Illumina/ExpansionHunter)
(selected fields from output JSON)
and the Read Alignments from [REViewer](https://github.com/Illumina/REViewer).
The loci were selected based on the variant catalog available at the ExpansionHunter
[GitHub repo](https://github.com/Illumina/ExpansionHunter/blob/c33a6faee33ed30557e3fa98170a81c65bf26d41/variant_catalog/hg38/variant_catalog.json).

```{r select_json_fields}
jsons <- dir_ls(params$indir, glob = "*.json", recurse = TRUE) |>
  as_tibble_col("fname") |>
  mutate(sample = basename(dirname(fname)))

pluck_json <- function(fname, sname) {
  pluck_locus_results <- function(j, sname) {
    sex <- j[["SampleParameters"]][["Sex"]]
    x <- j[["LocusResults"]]
    map(x, function(l) {
      list(
        Sex = sex,
        Locus = l$LocusId,
        ReferenceRegion = l$Variants[[1]][["ReferenceRegion"]],
        Repeat = l$Variants[[1]][["RepeatUnit"]],
        GT = l$Variants[[1]][["Genotype"]],
        Cov = l$Coverage,
        FragLen = l$FragmentLength
      )
    }) |>
      bind_rows() |>
      mutate(Sample = sname, Cov = floor(Cov)) |>
      select(Sample, everything())
  }

  read_json(fname) |>
    pluck_locus_results(sname) |>
    bind_rows()
}

x <- vector("list", length = nrow(jsons))
names(x) <- jsons$sample
for (i in seq_len(nrow(jsons))) {
  x[[jsons$sample[i]]] <- pluck_json(jsons$fname[i], jsons$sample[i])
}
d <- bind_rows(x)
```

```{r find_previewer_svg}
# find SVG files in indir
svg <- dir_ls(params$indir, glob = "*.svg", recurse = TRUE) |>
  as_tibble_col("fname") |>
  mutate(bname = basename(fname)) |>
  tidyr::separate(bname, into = c("sample", "locus", "svg"), sep = "\\.") |>
  select(-svg)

samples <- unique(svg$sample)
loci <- unique(svg$locus)
svg_sample_locus <- function(svg, s, l) {
  svg |> filter(sample == s, locus == l) |> pull(fname)
}
kable(as_tibble_col(samples, "samples")) |>
  kable_minimal()
```

## Results {.tabset .tabset-pills}

```{r show_previewer_svg, results='asis'}
cat('\n\n')
for (l in loci) {
  cat(glue("### {l}"), "\n")
  print(d |> filter(Locus == l) |> kable() |> kable_material())
  for (s in samples) {
    cat(glue("#### {s}"), "\n")
    cat(glue("![]({svg_sample_locus(svg, s, l)})"), "\n")
    cat('\n\n')
    cat('---')
    cat('\n\n')
  }
  cat('\n\n')
}
```
